pageserver/src/layered_repository.rs

```diff
@@ -355,13 +355,34 @@ impl LayeredRepository {
         timelineid: ZTimelineId,
         tenantid: ZTenantId,
         data: &TimelineMetadata,
+        first_save: bool,
     ) -> Result<PathBuf> {
-        let path = conf.timeline_path(&timelineid, &tenantid).join("metadata");
-        let mut file = File::create(&path)?;
+        let timeline_path = conf.timeline_path(&timelineid, &tenantid);
+        let path = timeline_path.join("metadata");
+        // use OpenOptions to ensure file presence is consistent with first_save
+        let mut file = OpenOptions::new()
```

Comment:
Truncating the file in place is scary. If you crash after the truncation, but before the new contents have been written out, you don't have any version of it.

In PostgreSQL, the 'control file' is somewhat similar to the metadata file here. It's created as a fixed-size file when the cluster is initialized, and it's overwritten on every checkpoint. It's small, less than 512 bytes, and we assume that it's written atomically. We could do the same here, but that requires that the file's size is never changed.

An alternative is to create the new metadata file with a temporary name, and rename() it in place.

Reply:
Is the rename crash safe?
Also would this be safe if I got rid of the truncate and always wrote a 512 byte vec (zeroes at the end)?

Reply:
> Is the rename crash safe?
I think the safe pattern is
1. create with temp name
2. fsync the file. (Or at least the contents, i.e. fdatasync or File::sync_data).
3. rename the file in place
4. fsync the file metadata (File::sync_all)

> Also would this be safe if I got rid of the truncate and always wrote a 512 byte vec (zeroes at the end)?

Yeah, that's acceptable. I don't think that guarantee is written down in any authoritative source like POSIX, but we assume that in Postgres with the control file and haven't had problems, and other software make similar assumptions.